### PR TITLE
Implement graceful exit

### PR DESCRIPTION
**Pull request by psibi**
The implementation mirrors somewhat what we do in the haskell binary pid1.

The earlier implementation had the disadvantage that if a SIGTERM or SIGINT handler was installed in the application which ignored the signals, it would not be terminated. While this may not be a problem in most of the case as the application would most likely be terminated by the cloud platform, it can be problematic in certain cases: https://stackoverflow.com/q/7743039/1651941

This PR makes sure that we send SIGKILL to the application and it's child processes. Unfortunately this is a breaking change.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -58,10 +97,28 @@ fn relaunch() -> Result<Child, Error> {
...
+        let _ = kill(Pid::from_raw(child_pid), Some(nix::sys::signal::SIGINT));
+        std::thread::sleep(settings.timeout);
+    }
+    // Send SIGTERM to all the processes with pid > 1
+    let _ = kill(Pid::from_raw(child_pid), Some(nix::sys::signal::SIGTERM));
+    std::thread::sleep(settings.timeout);
+
+    // Okay, some children are still present. Use the SIGKILL card.
+    let _ = kill(Pid::from_raw(child_pid), Some(nix::sys::signal::SIGKILL));
+    std::thread::sleep(settings.timeout);
```
Comment (snoyberg) [inline src/lib.rs]: Why sleep here? Shouldn't be necessary. If anything, I'd say sleep and then exit with an error code indicating that something impossible occurred.

File: src/lib.rs [hunk 2]
```diff
@@ -70,30 +127,11 @@ fn pid1_handling(settings: &Pid1Settings, child: Option<Child>) -> ! {
...
-                                }
-                                std::process::exit(exit_code)
-                            }
-                        }
-                    }
-                    None => std::process::exit(exit_code),
-                }
+                // We also do graceful exit in a separate thread so that
+                // pid1 exits as soon as possible
+                let _ = std::thread::spawn(move || gracefull_exit(settings, signal, child));
```
Comment (snoyberg) [inline src/lib.rs]: This has the potential to spawn this graceful exit thread twice. I'd mentioned the possibility of a `oneshot`, but a simpler solution may be sufficient: keeping a mutable variable in this function `let mut is_exiting = false§ and then, here, checking is `is_exiting` is false before spawning. When spawning you'd need to set `is_exiting` to true.

File: src/lib.rs (hunk 1 above)
Comment (psibi) [reply src/lib.rs]: You're right, sleep is not required here. I have removed it, thanks!

File: src/lib.rs (hunk 2 above)
Comment (psibi) [reply src/lib.rs]: Yeah, just like how you mentioned previously. :-) Thanks, have addressed it using an enum.

**Review by snoyberg — APPROVED** [review]

**Comment by snoyberg** [comment]
LGTM. This will be a nice improvement!

**Pull request merged**